lua/null-ls/builtins/formatting/nix_flake_fmt.lua

```diff
@@ -0,0 +1,154 @@
+local h = require("null-ls.helpers")
+local methods = require("null-ls.methods")
+local log = require("null-ls.logger")
+
+local FORMATTING = methods.internal.FORMATTING
+
+--- Return the command that `nix fmt` would run, or nil if we're not in a
+--- flake.
+---
+--- The formatter must follow treefmt's [formatter
+--- spec](https://github.com/numtide/treefmt/blob/main/docs/formatter-spec.md).
+---
+--- This basically re-implements the "entrypoint discovery" that `nix fmt` does.
+--- So why are we doing this ourselves rather than just invoking `nix fmt`?
+--- Unfortunately, it can take a few moments to evaluate all your nix code to
+--- figure out the formatter entrypoint. It can even be slow enough to exceed
+--- Neovim's default LSP timeout.
+--- By doing this ourselves, we can cache the result.
+---
+---@return string|nil
+local find_nix_fmt = function(params)
+    -- Discovering currentSystem here lets us keep the *next* eval pure.
+    -- We want to keep that part pure as a performance improvement: an impure
+    -- eval that references the flake would copy *all* files (including
+    -- gitignored files!), which can be quite expensive if you've got many GiB
+    -- of artifacts in the directory. This optimization can probably go away
+    -- once the [Lazy trees PR] lands.
+    --
+    -- [Lazy trees PR]: https://github.com/NixOS/nix/pull/6530
+    local cp = vim.system({ "nix", "--extra-experimental-features", "nix-command", "config", "show", "system" }):wait()
+    if cp.code ~= 0 then
+        log:warn(string.format("unable to discover builtins.currentSystem from nix. stderr: %s", cp.stderr))
+        return nil
+    end
+    local nix_current_system = cp.stdout:match("^(.-)%s+") -- remove trailing newline
+
+    local eval_nix_formatter = [[
+      let
+        currentSystem = "]] .. nix_current_system .. [[";
+        # Various functions vendored from nixpkgs lib (to avoid adding a
+        # dependency on nixpkgs).
+        lib = rec {
+          getOutput = output: pkg:
+            if ! pkg ? outputSpecified || ! pkg.outputSpecified
+            then pkg.${output} or pkg.out or pkg
+            else pkg;
+          getBin = getOutput "bin";
+          # Simplified by removing various type assertions.
+          getExe' = x: y: "${getBin x}/bin/${y}";
+          # getExe is simplified to assume meta.mainProgram is specified.
+          getExe = x: getExe' x x.meta.mainProgram;
+        };
+      in
+      formatterBySystem:
+        if formatterBySystem ? ${currentSystem} then
+          let
+            formatter = formatterBySystem.${currentSystem};
+            drv = formatter.drvPath;
+            bin = lib.getExe formatter;
+          in
+            drv + "\n" + bin + "\n"
+        else
+          ""
+    ]]
+
+    cp = vim.system({
+        "nix",
+        "--extra-experimental-features",
+        "nix-command flakes",
+        "eval",
+        ".#formatter",
```

Comment:
Another question is, how do you invalidate the cache if the flake changes? With this changes are only ever applied if you restart vim.

Reply:
In nix-direnv we have some heuristic for this.

Reply:
> Another question is, how do you invalidate the cache if the flake changes? With this changes are only ever applied if you restart vim

Good question. I was hoping to avoid cache-invalidation heuristics here. My thinking:

- Somewhat lame justification: the caching story here is no worse (and arguably better) than the approach in this PR: https://github.com/nvimtools/none-ls.nvim/pull/189, which would require reloading your nix devshell and then restarting Neovim with an updated `PATH` env var
- Less lame: I was thinking we could make it easy for people to reset the cache without restarting neovim. More on that below.

There is this undocumented, possibly-only-for-testing cache reset functionality:

`:lua require("null-ls.helpers").cache._reset()`

It's kind of buggy. I fixed that in https://github.com/nvimtools/none-ls.nvim/pull/199/files.

<details>
<summary>I had an idea here that didn't pan out. You don't need to read this. I just wanted to write down my discovery someplace in case it's relevant in the future.</summary>

My *hope* was that people could just clear the cache by "restarting" the LSP client, but I now realize that restarting a LSP client isn't really a thing in neovim (yet). You can stop a client, but there's no general mechanism for starting them back up. There's `:LspRestart` from lspconfig, and there's some discussion about upstreaming that into neovim core over here: https://github.com/neovim/neovim/issues/28479 (which would require neovim core to gain the concept of "registering" LSP clients separate from starting them. In other words, there's no good way to restart the none-ls LSP client: you can stop it, but then there's [idepotency logic](https://github.com/nvimtools/none-ls.nvim/blob/4b3ce339e7918d97c0b7ee27a37c832950f95547/lua/null-ls/config.lua#L92-L94) that prevents you from starting it back up.
</details>

> In nix-direnv we have some heuristic for this. A minimum I would check the modification date of flake.nix and flake.lock.

So I started building this, and then realized that it's pretty annoying when you're editing a `flake.nix` file, as you're just going to bust the cache every single time you change the file. I know the implementation here is imperfect, but I believe the caching we already have in this PR + the ability to bust the cache (either by restarting neovim or by calling the cache reset function I mentioned above) is a better solution.

Reply:
I'm marking this resolved. Feel free to reopen if you'd like a change here!